[Mail compose] Color picker isn't shown on macOS Ventura via context menu -> Font -> Show Colors

whsieh opened the pull request:
#### 38a346199d28cd17a706767e09a9af94f7ed996a
<pre>
[Mail compose] Color picker isn&apos;t shown on macOS Ventura via context menu -&gt; Font -&gt; Show Colors
<a href="https://bugs.webkit.org/show_bug.cgi?id=242557">https://bugs.webkit.org/show_bug.cgi?id=242557</a>
rdar://95659711

Reviewed by Devin Rousso.

In macOS Ventura, Mail now uses WebKit2 for its editable compose web view. However, in WebKit2,
we&apos;ve had a long-standing bug where the Font, Style, and Color menu items (all under the Font menu)
don&apos;t work properly, since they still attempt to display their respective system-wide editing
controls by calling into the shared `NSFontManager` or `NSApplication` from within the web process
(in `Editor` code).

This is a consequence of the items being handled by `ContextMenuController` in the web process,
which is normal for the vast majority of other menu items, such as &quot;Copy&quot; and &quot;Search with …&quot;, but
is currently incompatible with these items. There are two general approaches we can take to fix this
issue:

1.  Add new editor (or chrome) client hooks to delegate showing the font, color and style panel out
    to the client layer, in WebKit1 and WebKit2.

2.  Handle the action in the client layer in the first place, avoiding the need to IPC back to the
    web process (or in the case of WebKit1, call back into WebCore).

I opted for (2) in this patch, since it ends up simplifying this code (and skips a bit of extra IPC
messaging in WebKit2). See below for more details.

Test: ContextMenuTests.ShowColorPanel

* LayoutTests/editing/mac/style/do-not-crash-when-showing-font-panel-expected.txt: Removed.
* LayoutTests/editing/mac/style/do-not-crash-when-showing-font-panel.html: Removed.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Remove this layout test, since it&apos;s superceded by the new API test, which actually verifies that the
right platform API is invoked when selecting one of the context menu items. In its current form,
this test now just hits a debug assertion after this patch, since we don&apos;t expect
`ContextMenuController` to handle the font, color and style panel items.

* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::showFontPanel): Deleted.
(WebCore::Editor::showStylesPanel): Deleted.
(WebCore::Editor::showColorPanel): Deleted.
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::showFontPanel): Deleted.
(WebCore::Editor::showStylesPanel): Deleted.
(WebCore::Editor::showColorPanel): Deleted.

Remove these macOS-specific methods from `Editor` entirely, now that they&apos;re no longer needed.

* Source/WebCore/loader/EmptyClients.cpp:

Revert the `canShowFontPanel()` client hook added in `213881@main`, since it&apos;s no longer necessary
after adding support for this in WebKit2.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
* Source/WebCore/page/EditorClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:

Immediately handle showing the font, color and style panels in the client layer, instead of calling
back into the web process. See below.

(WebKit::WebPageProxy::contextMenuItemSelected):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showFontPanel):
(WebKit::WebPageProxy::showStylesPanel):
(WebKit::WebPageProxy::showColorPanel):

Add macOS-only helper methods to show the font, styles and color panels. Note that for this patch,
I&apos;m opting to limit this to editable web views for the time being to avoid exposing the new editing
commands made possible by the font panel (e.g. the ability to adjust font shadows, font family, and
other attributes) to the web. We should remove these checks (and thereby fully fix the long-standing
bug rdar://21577518) in a followup, when we have more time to assess the developer impact of
enabling these capabilities.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebMenuTarget forwardContextMenuAction:]):
* Source/WebKitLegacy/win/WebCoreSupport/WebEditorClient.h:
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(itemMatchingFilter):
(-[TestWKWebView rightClick:andSelectItemMatching:]):

Add an API test to synthesize a right click in an editable web view, and select the &quot;Show Colors&quot;
context menu action underneath the fonts menu. To achieve this, we factor out some existing logic
for triggering a right click and selecting a menu item from the resulting context menu into a
separate category helper method, `-rightClick:andSelectItemMatching:`. We also teach this method to
recursively traverse through submenus in the given `NSMenu` (in order to actually find and activate
the &quot;Show Colors&quot; item, which is underneath the &quot;Fonts&quot; submenu). This test swizzles the method for
showing the color panel, `-[NSApplication orderFrontColorPanel:]`, and ends successfully when this
is called after activating the menu item.

(TestWebKitAPI::swizzledOrderFrontColorPanel):

Canonical link: <a href="https://commits.webkit.org/252345@main">https://commits.webkit.org/252345@main</a>
</pre>

dcrousso commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -536,6 +525,9 @@ void ContextMenuController::contextMenuItemSelected(ContextMenuAction action, co
     case ContextMenuItemTagDictationAlternative:
         frame->editor().applyDictationAlternative(title);
         break;
+    case ContextMenuItemTagShowFonts:
+    case ContextMenuItemTagStyles:
+    case ContextMenuItemTagShowColors:
```
Will these now `ASSERT_NOT_REACHED` on other platforms?  Should we wrap them in a `#if PLATFORM(MAC)`?

whsieh replied:
Yes (though, there shouldn't be any way to trigger these on non-Mac platforms so that's expected).

dcrousso reviewed (approved):
oops in commit message
> is currently incompatible with the , there are two general approaches we can take:

whsieh committed:
[Mail compose] Color picker isn't shown on macOS Ventura via context menu -> Font -> Show Colors
https://bugs.webkit.org/show_bug.cgi?id=242557
rdar://95659711

Reviewed by Devin Rousso.

In macOS Ventura, Mail now uses WebKit2 for its editable compose web view. However, in WebKit2,
we've had a long-standing bug where the Font, Style, and Color menu items (all under the Font menu)
don't work properly, since they still attempt to display their respective system-wide editing
controls by calling into the shared `NSFontManager` or `NSApplication` from within the web process
(in `Editor` code).

This is a consequence of the items being handled by `ContextMenuController` in the web process,
which is normal for the vast majority of other menu items, such as "Copy" and "Search with …", but
is currently incompatible with these items. There are two general approaches we can take to fix this
issue:

1.  Add new editor (or chrome) client hooks to delegate showing the font, color and style panel out
    to the client layer, in WebKit1 and WebKit2.

2.  Handle the action in the client layer in the first place, avoiding the need to IPC back to the
    web process (or in the case of WebKit1, call back into WebCore).

I opted for (2) in this patch, since it ends up simplifying this code (and skips a bit of extra IPC
messaging in WebKit2). See below for more details.

Test: ContextMenuTests.ShowColorPanel

* LayoutTests/editing/mac/style/do-not-crash-when-showing-font-panel-expected.txt: Removed.
* LayoutTests/editing/mac/style/do-not-crash-when-showing-font-panel.html: Removed.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Remove this layout test, since it's superceded by the new API test, which actually verifies that the
right platform API is invoked when selecting one of the context menu items. In its current form,
this test now just hits a debug assertion after this patch, since we don't expect
`ContextMenuController` to handle the font, color and style panel items.

* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::showFontPanel): Deleted.
(WebCore::Editor::showStylesPanel): Deleted.
(WebCore::Editor::showColorPanel): Deleted.
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::showFontPanel): Deleted.
(WebCore::Editor::showStylesPanel): Deleted.
(WebCore::Editor::showColorPanel): Deleted.

Remove these macOS-specific methods from `Editor` entirely, now that they're no longer needed.

* Source/WebCore/loader/EmptyClients.cpp:

Revert the `canShowFontPanel()` client hook added in `213881@main`, since it's no longer necessary
after adding support for this in WebKit2.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
* Source/WebCore/page/EditorClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:

Immediately handle showing the font, color and style panels in the client layer, instead of calling
back into the web process. See below.

(WebKit::WebPageProxy::contextMenuItemSelected):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showFontPanel):
(WebKit::WebPageProxy::showStylesPanel):
(WebKit::WebPageProxy::showColorPanel):

Add macOS-only helper methods to show the font, styles and color panels. Note that for this patch,
I'm opting to limit this to editable web views for the time being to avoid exposing the new editing
commands made possible by the font panel (e.g. the ability to adjust font shadows, font family, and
other attributes) to the web. We should remove these checks (and thereby fully fix the long-standing
bug rdar://21577518) in a followup, when we have more time to assess the developer impact of
enabling these capabilities.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebMenuTarget forwardContextMenuAction:]):
* Source/WebKitLegacy/win/WebCoreSupport/WebEditorClient.h:
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(itemMatchingFilter):
(-[TestWKWebView rightClick:andSelectItemMatching:]):

Add an API test to synthesize a right click in an editable web view, and select the "Show Colors"
context menu action underneath the fonts menu. To achieve this, we factor out some existing logic
for triggering a right click and selecting a menu item from the resulting context menu into a
separate category helper method, `-rightClick:andSelectItemMatching:`. We also teach this method to
recursively traverse through submenus in the given `NSMenu` (in order to actually find and activate
the "Show Colors" item, which is underneath the "Fonts" submenu). This test swizzles the method for
showing the color panel, `-[NSApplication orderFrontColorPanel:]`, and ends successfully when this
is called after activating the menu item.

(TestWebKitAPI::swizzledOrderFrontColorPanel):

Canonical link: https://commits.webkit.org/252345@main
